File: content/14-2d-dynamic-programming/partition_equal_subset_sum.py
def canPartition(nums) -> bool:
    total = sum(nums)
    if total % 2 != 0:
        return False

    target = total // 2
    dp = set([0])

    for n in nums:
        nextDP = set(dp)
        for t in dp:
            if t + n == target:
                return True
            if t + n < target:
                nextDP.add(t + n)
        dp = nextDP

    return target in dp

if __name__ == "__main__":
    assert canPartition([1,5,11,5]) == True, "Test 1 Failed"
    assert canPartition([1,2,3,5]) == False, "Test 2 Failed"
    print("All tests passed!")

File: content/14-2d-dynamic-programming/partition_equal_subset_sum.js
function canPartition(nums) {
    let total = nums.reduce((a, b) => a + b, 0);
    if (total % 2 !== 0) return false;

    let target = total / 2;
    let dp = new Set([0]);

    for (let n of nums) {
        let nextDP = new Set(dp);
        for (let t of dp) {
            if (t + n === target) return true;
            if (t + n < target) nextDP.add(t + n);
        }
        dp = nextDP;
    }

    return dp.has(target);
}

const assert = require('assert');
assert.strictEqual(canPartition([1,5,11,5]), true);
assert.strictEqual(canPartition([1,2,3,5]), false);
console.log("All tests passed!");
